photobpf.c

```diff
@@ -0,0 +1,275 @@
+/*
+** ATOP - System & Process Monitor
+**
+** The program 'atop' offers the possibility to view the activity of
+** the system on system-level as well as process-level.
+**
+** This source-file contains functions to read bpf-porgram counters.
+** ================================================================
+** Author:      Song Liu
+** E-mail:      song@kernel.org
+** Date:        Dec 2019
+**
+** This program is free software; you can redistribute it and/or modify it
+** under the terms of the GNU General Public License as published by the
+** Free Software Foundation; either version 2, or (at your option) any
+** later version.
+**
+** This program is distributed in the hope that it will be useful, but
+** WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+** See the GNU General Public License for more details.
+*/
+
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <string.h>
+#include <curses.h>
+#include <regex.h>
+
+#include "atop.h"
+#include "photobpf.h"
+#include "showgeneric.h"
+
+#define BPF_STATS_MIN_ALLOC 8
+
+/*
+** check configurations:
+**   bpfsampleinterval < interval
+*/
+void
+photo_bpf_check(void)
+{
+	if (bpfsampleinterval >= interval)
+		mcleanstop(1, "bpfsampleinterval must be smaller than interval");
+}
+
+/* boolean */
+int
+system_support_bpf(void)
+{
+	int bpf_stats_fd = bpf_enable_stats(BPF_STATS_RUN_TIME);
+
+	/* check bpf_enable_stats() returns valid fd */
+	if (bpf_stats_fd < 0)
+		return 0;
+	close(bpf_stats_fd);
+	return 1;
+}
+
+/*
+** Get a snapshot of all bpf program stats
+*/
+static struct bstats *
+get_allbstats(void)
+{
+	int		err;
+	__u32		id = 0;
+	unsigned int	bufsize = 0;
+	struct bstats	*curbstats = malloc(sizeof(struct bstats));
+
+	ptrverify(curbstats, "Malloc failed for bstats\n");
+
+	memset(curbstats, 0, sizeof(struct bstats));
+
+	while (true) {
+		int			fd;
+		struct bpf_prog_info	info = {};
+		__u32			len = sizeof(struct bpf_prog_info);
+		struct bstat		*bstat;
+
+		err = bpf_prog_get_next_id(id, &id);
+		if (err)
+			break;
+		if (curbstats->nbpfall >= bufsize) {
+			bufsize += BPF_STATS_MIN_ALLOC;
+			curbstats->bpfall = realloc(
+				curbstats->bpfall,
+				bufsize * sizeof(struct bstat));
+			ptrverify(curbstats->bpfall,
+				  "Malloc failed for %u bstats\n",
+				  bufsize);
+		}
+
+		fd = bpf_prog_get_fd_by_id(id);
+		if (fd < 0 && errno != ENOENT)
+			break;
+
+		err = bpf_obj_get_info_by_fd(fd, &info, &len);
+		if (err) {
+			close(fd);
+			break;
+		}
```

Comment:
Aren't you leaking `fd` if it doesn't go on the error path?

Reply:
Good catch! Fixing it.